Sources/Plasma/NucleusLib/pnAsyncCoreExe/pnAceDns.cpp

```diff
@@ -0,0 +1,155 @@
+permission to convey the resulting work. Corresponding Source for a
+non-source form of such a combination shall include the source code for
+the parts of OpenSSL and IJG JPEG Library used as well as that of the covered
+work.
+
+You can contact Cyan Worlds, Inc. by email legal@cyan.com
+ or by snail mail at:
+      Cyan Worlds, Inc.
+      14617 N Newport Hwy
+      Mead, WA   99021
+
+*==LICENSE==*/
+
+#include "Pch.h"
+
+using tcp = asio::ip::tcp;
+
+struct DnsResolver
+{
+    asio::io_context m_context;
+    asio::executor_work_guard<asio::io_context::executor_type> m_workGuard;
+    tcp::resolver m_resolver;
+    std::thread m_lookupThread;
+
+    DnsResolver()
+        : m_context(), m_resolver(m_context), m_workGuard(m_context.get_executor())
+    {
+        // Start the resolver thread
+        m_lookupThread = std::thread([this] { m_context.run(); });
+    }
+
+    void destroy(unsigned exitThreadWaitMs)
+    {
+        m_workGuard.reset();
+        m_resolver.cancel();
+        AsyncThreadTimedJoin(m_lookupThread, exitThreadWaitMs);
+        m_lookupThread = {};
+    }
+};
+
+struct DnsResolveData
+{
+    ST::string          m_name;
+    FAsyncLookupProc    m_lookupProc;
+    void*               m_param;
+};
+
+static std::recursive_mutex     s_critsect;
+static DnsResolver             *s_resolver = nullptr;
+
+void DnsDestroy(unsigned exitThreadWaitMs)
+{
+    if (s_resolver) {
+        s_resolver->destroy(exitThreadWaitMs);
+        delete s_resolver;
+        s_resolver = nullptr;
+    }
+}
+
+static void AddressResolved(const asio::error_code& err,
+                            const tcp::resolver::results_type& results,
+                            const DnsResolveData& data)
+{
+    if (err || results.empty()) {
+        if (err)
+            LogMsg(kLogFatal, "DNS: Failed to resolve {}: {}", data.m_name, err.message());
+        data.m_lookupProc(data.m_param, data.m_name, 0, nullptr);
+        return;
+    }
+
+    // This can be cleaned up when the rest of the netcode uses ASIO instead
+    // of other (Win32/NT) backends...
+    std::vector<plNetAddress> addrs;
+    addrs.reserve(results.size());
+    for (const auto& result : results) {
+        const auto& endpoint = result.endpoint();
+        if (!endpoint.address().is_v4()) {
+            // Due to limitations in the net protocol, we can only support IPv4 for now :(
+            continue;
+        }
+        const auto ipv4_addr = endpoint.address().to_v4();
+        addrs.emplace_back(ipv4_addr.to_bytes(), endpoint.port());
+    }
+
+    data.m_lookupProc(data.m_param, data.m_name, addrs.size(), addrs.data());
+
+    PerfSubCounter(kAsyncPerfNameLookupAttemptsCurr, 1);
+}
+
+void AsyncAddressLookupName(FAsyncLookupProc lookupProc,
+                            const ST::string& name, unsigned port, void* param)
+{
+    ASSERT(lookupProc);
+
+    PerfAddCounter(kAsyncPerfNameLookupAttemptsCurr, 1);
+    PerfAddCounter(kAsyncPerfNameLookupAttemptsTotal, 1);

```

Comment:
Is it possible to use plProfile for these counters? Maybe not without making plProfile thread-safe and atomic?

Reply:
I used these because they were already in place from the old DNS code...  As you mentioned, plProfile is not atomic, which would be a problem for these counters with the current design, so I feel like that sort of change belongs in a separate PR.

Reply:
I'd prefer for us to avoid making `plProfile` atomic if at all possible.